03-attribute-operations.Rmd

```diff
@@ -499,14 +507,211 @@ world_data = world %>% st_set_geometry(NULL)
+It ranges between - 1 and 1 with positive values indicating the presence of living plants (mostly > 0.2).
+To calculate the NDVI, one uses the red and near-infrared bands of remotely sensed imagery (e.g., Landsat or Sentinel imagery) exploiting the fact that vegetation absorpts light heavily in the visible light spectrum, and especially in the red channel, while reflecting it in the near-infrared spectrum.
+
+$$
+\begin{split}
+NDVI&= \frac{\text{NIR} - \text{Red}}{\text{NIR} + \text{Red}}\\
+\end{split}
+$$
+where NIR = near infrared channel
+      Red = red channel
+
+Predictive mapping is another interesting application of local raster operations.
+The response variable correspond to measured or observed points in space, for example, species richness, the presence of landslides, tree disease or crop yield.
+Consequently, we can easily retrieve space- or airborne predictor variables from various rasters (elevation, pH, precipitation, temperature, landcover, soil class, etc.).
+Subsequently, we model our response as a function of our predictors using `lm`, `glm`, `gam` or a machine-learning technique. 
+To make a spatial prediction, all we have to do, is to apply the estimated coefficients to the predictor rasters, and summing up the resulting output rasters (Chapter ??; see also @muenchow_predictive_2013).
+<!-- add reference to chapter ecological modeling -->
+
+While local functions operate on one cell, though possibly from multiple layers, **focal** operations take into account a central cell and its neigbors.
+The neigborhood (also named kernel, filter or moving window) under consideration is typically of size 3-by-3 cells (that is the central cell and its eight surrounding neighbors) but can take on any other (not necessarily rectangular) shape as defined by the user.
+A focal operation applies an aggregation function to all cells within the specified neighborhood, uses the corresponding output as the new value for the the central cell, and moves on to the next central cell (Figure @ref(fig:focal_example)).
+Other names for this operation are spatial filtering and convolution [@burrough_principles_2015].
+
+In R, we can use the `focal()` function to perform spatial filtering. 
+We define the shape of the moving window with a `matrix` whose values correspond to weights.
+Secondly, the the `fun` argument lets us specify the function we wish to apply to this neighborhood.
+Here, we choose the minimum, but of course we can use any other function such as the the sum, the mean, the median, the mode, the maximum or the variance.
+
+```{r, eval = FALSE}
+r_focal = focal(r, w = matrix(1, nrow = 3, ncol = 3), fun = min)
+```
+
+```{r focal_example, echo = FALSE, fig.cap = "Input raster (left) and resulting output raster (right) due to a focal operation - summing up 3-by-3 windows"}
+knitr::include_graphics("figures/03_focal_example.png")
```

Comment:
This is an excellent figure, great work.

Reply:
Thanks, glad you like it!